bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/ADBridgeHandler.java

```diff
@@ -0,0 +1,428 @@
+     * @throws MessageParseException
+     */
+    private void parseLRRMessage(String msg) throws MessageParseException {
+        LRRMessage lrrm;
+
+        // Parse the message
+        try {
+            lrrm = new LRRMessage(msg);
+        } catch (IllegalArgumentException e) {
+            throw new MessageParseException(e.getMessage());
+        }
+
+        // Notify appropriate LRRHandlers
+        Collection<LRRHandler> handlers = findLRRHandlers(lrrm.partition);
+        for (LRRHandler handler : handlers) {
+            handler.handleUpdate(lrrm);
+        }
+    }
+
+    /**
+     * Parse and handle version (VER) message. This just updates bridge properties.
+     *
+     * @param msg string containing incoming message payload
+     * @throws MessageParseException
+     */
+    private void parseVersionMessage(String msg) throws MessageParseException {
+        VersionMessage verm;
+
+        try {
+            verm = new VersionMessage(msg);
+        } catch (IllegalArgumentException e) {
+            throw new MessageParseException(e.getMessage());
+        }
+
+        logger.trace("Processing version message sn:{} ver:{} cap:{}", verm.serial, verm.version, verm.capabilities);
+        Map<String, String> properties = editProperties();
+        properties.put(PROPERTY_SERIALNUM, verm.serial);
+        properties.put(PROPERTY_VERSION, verm.version);
+        properties.put(PROPERTY_CAPABILITIES, verm.capabilities);
+        updateProperties(properties);
+    }
+
+    /**
+     * Notify all child thing handlers that the alarm panel is in the ready state. Since there is no way to poll, all
+     * channels are initialized into the UNDEF state. This method is called when there is reason to assume that there
+     * are no faulted zones, because the alarm panel is in state READY. Zone handlers that have not yet received updates
+     * can then set their contact states to CLOSED. Only executes the first time panel is ready after bridge
+     * connect/reconnect.
+     */
+    private void notifyChildHandlersPanelReady() {
+        if (!panelReadyReceived) {
+            panelReadyReceived = true;
+            logger.trace("Notifying child handlers that panel is in ready state");
+
+            // Notify child zone handlers by calling notifyPanelReady() for each
+            for (Thing thing : getThing().getThings()) {
+                if (thing.getHandler() instanceof ZoneHandler || thing.getHandler() instanceof RFZoneHandler) {
+                    ADThingHandler handler = (ADThingHandler) thing.getHandler();
+                    if (handler != null) {
+                        handler.notifyPanelReady();
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Return the ZoneHandler for the given address and channel, or null if no handler exists and is initialized.
+     */
+    private @Nullable ZoneHandler findZoneHandler(int address, int channel) {
```

Comment:
Right now you have this designed such that the bridge searches out handlers that should process a given message. This requires the bridge to have knowledge of all of the ADHandler subclasses and also a two step approach:
1. Bridge must identify which handlers could handle a message.
2. Bridge then passes on message to handlers found in step 1

I think it would be simpler if instead you just send the message to all of the handlers and let the handlers decide whether or not they want to handle it. This allows the bridge to be ignorant of the types of handlers as well as possibly allowing the bridge to be ignorant of the message type as well. Also instead of each handler having to implement 2 methods (one to judge if it can handle the message and another to actually handle it), each handler would only have to implement a single method which optionally handles the message.

Reply:
Yeah, I'll admit I don't love the way this works at the moment. I'm not sure I really like the idea of sending every message to every handler either, though, although it would probably eliminate a lot of code.  The AD protocol can be very chatty, especially with keypad messages which it seems to send every few seconds. I'd rather not hit every zone/rfzone handler with every keypad message, since there can be a lot of both. Sending each message to every handler of the correct type may be ok, but that would still require maintaining or generating a list of the child handlers of each type. Originally I was planning on registering all of the child handlers in maps in the bridge so that look-ups would be very efficient, but that seemed complex and error-prone, especially given the address mask comparisons necessary for finding the appropriate keypad or lrr handler for a given message.
Given a choice, I would rather not mess with this for now, since it works well enough, and come up with a better scheme later.

Reply:
Ok. I had some time today, so I rewrote this part of the code. The bridge now sends messages to all child handlers of the correct type, and they decide in their handleUpdate(ADMessage) methods whether or not they should act on them. I'll merge it in to this branch after a bit more testing.